nemo/collections/asr/parts/k2/classes.py

```diff
@@ -0,0 +1,165 @@
+        # use k2 import guard
+        k2_import_guard()
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.graph_module_cfg = self._cfg.graph_module_cfg
+
+        # register token_lm for MAPLoss
+        criterion_type = self.graph_module_cfg.get("criterion_type", "ml")
+        self.use_graph_lm = criterion_type == "map"
+        if self.use_graph_lm:
+            token_lm_path = self.graph_module_cfg.background_cfg.get("token_lm", None)
+            if token_lm_path is None:
+                raise ValueError(
+                    f"graph_module_cfg.background_cfg.token_lm is empty. It must be set for criterion_type == `{criterion_type}`"
+                )
+            token_lm_path = self.register_artifact('graph_module_cfg.background_cfg.token_lm', token_lm_path)
+            self.graph_module_cfg.background_cfg["token_lm"] = token_lm_path
+
+        self.update_k2_modules(self.graph_module_cfg)
+
+    def update_k2_modules(self, input_cfg):
+        """
+        Helper function to initialize or update k2 loss and transcribe_decoder.
+        """
+        del self.loss
+        if hasattr(self, "transcribe_decoder"):
+            del self.transcribe_decoder
+
+        self.loss = LatticeLoss(
+            num_classes=self.decoder.num_classes_with_blank - 1,
+            reduction=self._cfg.get("ctc_reduction", "mean_batch"),
+            backend="k2",
+            criterion_type=input_cfg.get("criterion_type", "ml"),
+            split_batch_size=input_cfg.get("split_batch_size", 0),
+            graph_module_cfg=input_cfg.background_cfg,
+        )
+        remove_consecutive = input_cfg.background_cfg.get(
+            "topo_with_self_loops", True
+        ) and input_cfg.background_cfg.get("topo_type", "default") not in ["forced_blank", "identity",]
+        self._wer.remove_consecutive = remove_consecutive
+
+        criterion_type = self.loss.criterion_type
+        self.use_graph_lm = criterion_type == "map"
+        transcribe_training = input_cfg.get("transcribe_training", False)
+        if transcribe_training and criterion_type == "ml":
+            logging.warning(
+                f"""You do not need to use transcribe_training=`{transcribe_training}` 
+                            with criterion_type=`{criterion_type}`. transcribe_training will be set to False."""
+            )
+            transcribe_training = False
+        self.transcribe_training = transcribe_training
+        if self.use_graph_lm:
+            self.transcribe_decoder = ViterbiDecoderWithGraph(
+                num_classes=self.decoder.num_classes_with_blank - 1,
+                backend="k2",
+                dec_type="tokenlm",
+                return_type="1best",
+                return_ilabels=True,
+                output_aligned=True,
+                split_batch_size=input_cfg.get("split_batch_size", 0),
+                graph_module_cfg=input_cfg.background_cfg,
+            )
+
+    @typecheck()
+    def forward(
```

Comment:
Do not override any functions of the base class that implements mixins, subclasses that inherit mixins must properly override their own functions and call appropriate mixin methods as needed.

Reply:
Fixed